namer/namer.cc

```diff
@@ -473,9 +481,44 @@ class SymbolFinder {
         return foundRef;
     }
 
+    // Returns `true` if `asgn` is a field declaration.
+    bool handleFieldDeclaration(core::Context ctx, ast::Assign &asgn) {
+        auto *uid = ast::cast_tree<ast::UnresolvedIdent>(asgn.lhs);
+        if (uid == nullptr) {
+            return false;
+        }
+
+        if (uid->kind != ast::UnresolvedIdent::Kind::Instance && uid->kind != ast::UnresolvedIdent::Kind::Class) {
+            return false;
+        }
+
+        auto *cast = ast::cast_tree<ast::Cast>(asgn.rhs);
+        if (cast == nullptr) {
+            return false;
+        }
+
+        // This is an error in resolver, but just ignore this here.
+        if (cast->cast != core::Names::let()) {
+            return false;
+        }
+
+        core::FoundField found;
+        found.kind = uid->kind == ast::UnresolvedIdent::Kind::Instance ? core::FoundField::Kind::Instance
+                                                                       : core::FoundField::Kind::Class;
+        found.owner = getOwner();
```

Comment:
I wonder if it makes sense to say here that the owner is actually the class that will own the field/static-field, and not have to bring methods into the fold?

e.g., if we stored (1) whether it was a class var or instance var and (2) whether it appeared in a self method, is that enough? (i'm not sure)

I'm asking because

- The logic around deleting methods gets hairy, and I would love for it to be completely obvious that method symbols can be added/deleted without touching instance variables
- It's somewhat odd that we do `addField` for something that we know will create a `static-field` symbol eventually even though we have a separate `FoundStaticField`. (I mostly see the line of reasoning how we arrived there.)

Maybe if added a `FoundStaticField` for the class vars, then the only bit we would need is whether the assignment happened in a singleton class method or instance method, and then the `found.owner` could be the enclosing class 100% of the time?

This is not necessarily blocking feedback, but given that it seems like `insertField` is never called, and calling it correctly involves intermixing logic across when the methods get deleted and re-added, I wonder if this approach sidesteps that complication entirely.

Reply:
I think I initially tried to get this all to work without pushing methods onto the ownership stack, and it got pretty hairy, as well as not being correct in all cases (some of that might have been the issues with alias methods that I didn't see until later?).  Note that you also need to handle variables appearing outside method definitions (the AST does not yet have explicit `static-init` methods at this point).

I could see that you could have `FoundDefinition::Kind::MethodRef` or something on the ownership stack representing that you are now inside a method, but computing the owner of a `MethodRef` in `getOwnerSymbol` doesn't involve looking at `definedMethods`.